subsys/net/lib/download_client/src/download_client.c

```diff
@@ -655,8 +672,21 @@ int download_client_disconnect(struct download_client *const client)
 	return 0;
 }
 
-int download_client_start(struct download_client *client, const char *file,
-			  size_t from)
+int download_client_cancel(struct download_client *const client)
+{
+	if (client == NULL || client->fd < 0) {
+		return -EINVAL;
+	}
+
+	is_cancelled = true;
+
+	/* wakeup thread to cancel the download */
+	k_thread_resume(client->tid);
```

Comment:
What if the socket has no timeout configured? `recv()` is a blocking call, therefore `download_client_cancel()` might be ineffective here.

Reply:
I agree. `download_client_cancel` the way I have implemented requires a timeout at some point when calling `recv()`. I just have reproduced the problem using download sample within sdk-nrf. The changes are below. I see two options here:
- fix the socket api in a way it can handle multi threading (should be done anway)
- honestly, I do not think that a download where recv might block forever is really practicable but I am not sure if you are willing to accept API changes. I would disallow setting timeout to 0. 
What are your thoughts?

```
diff --git a/samples/nrf9160/download/src/main.c b/samples/nrf9160/download/src/main.c
index 53bbdf39c..135d23ba8 100644
--- a/samples/nrf9160/download/src/main.c
+++ b/samples/nrf9160/download/src/main.c
@@ -165,6 +165,21 @@ static int callback(const struct download_client_evt *event)
        return 0;
 }
 
+void my_work_handler(struct k_work *work)
+{
+    printk("*********** cancelling\n");
+    download_client_disconnect(&downloader);
+}
+
+K_WORK_DEFINE(my_work, my_work_handler);
+
+void my_timer_handler(struct k_timer *dummy)
+{
+       k_work_submit(&my_work);
+}
+
+K_TIMER_DEFINE(my_timer, my_timer_handler, NULL);
+
 void main(void)
 {
        int err;
@@ -214,4 +229,6 @@ void main(void)
        }
 
        printk("Downloading %s\n", URL);
+
+       k_timer_start(&my_timer, K_SECONDS(5), K_NO_WAIT);
 }
 ```

Reply:
@rlubos: A few days have passed and I´ve come to conclusion that I would go with that:

- disallow blocking socket for download_client. Honestly, I do not see any reason to use blocking sockets in a download client. Even a timeout of several hours is better than no timeout.
- close socket within download client thread. This is clean, respects the caller-chain (who does what) and avoids unexpected behavior when dealing with fds (although I currently see no problems with the current implementation)

My goal is a deterministic behavior of the fota process i.e. once it is started, what ever happens, after some time it returns successfully or with an error and **never** gets stuck. This is very important for embedded devices that are not on your desk you can reboot easily at any time. I do not want to wrap a timer around the whole fota process to ensure this. This is not a nice coding style.

While searching for possible bugs I have that the method "request_send" within download_client.c might get stuck. I would like to introduce SO_SNDTIMEO sockopt for this as already done for SO_RCVTIMEO. What do you think about this?

Reply:
> disallow blocking socket for download_client.

Can you clarify what do you mean? I.e. do you want to rewrite the codebase to use non-blocking sockets or did you mean to require to set SO_RCVTIMEO in all cases?

> I would like to introduce SO_SNDTIMEO sockopt for this as already done for SO_RCVTIMEO

Sounds fine to me. @lemrey What do you think, any thoughts?

Reply:
> > disallow blocking socket for download_client.
> 
> Can you clarify what do you mean? I.e. do you want to rewrite the codebase to use non-blocking sockets or did you mean to require to set SO_RCVTIMEO in all cases?

Yes, I would require SO_RCVTIMEO and SO_SNDTIMEO to be set all the time with value > 0.

Reply:
@rlubos: any update on this?

Reply:
@lemrey I'd like to hear your opinion, you're the library author after all.

Reply:
It's reasonable to use a timeout for both `recv()` and `send()`. New libmodem will have better multi-threading support, it'll be possible to `close()` a socket from another thread and unblock other threads.

Reply:
I have implemented a proper timeout handling in an other PR.
https://github.com/nrfconnect/sdk-nrf/pull/7167
Once this PR is resolved, I suggest to take again this PR into a discussion

Reply:
Hi @andreas-chmielewski-gcx, the issue related to closing a socket from a thread when another thread is in a `recv()` call on the same socket should be resolved with libmodem v2.0.0.

Reply:
Hi @andreas-chmielewski-gcx, libmodem v2.0.0 was merged.